### PR TITLE
New component: Docker receiver

### PR DESCRIPTION
**Pull request by aboguszewski-sumo**
**Description:** This PR adds a new Docker receiver that is going to replace [Docker Stats receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/dockerstatsreceiver). 

Full list of changes between Docker stats receiver and Docker receiver (as of now):

- The name ;) 
- Config has undergone some changes: `stats` key was added for options that were related only to fetching stats. 
- I have added myself as a codeowner, alongside the codeowners for Docker stats receiver

I agree that this PR is quite big, but it's mostly generated or copied code.

The receiver is going to eventually support events and container logs, they will be added in separate pull requests.

I wasn't sure if I should deprecate dockerstatsreceiver in this PR or in a separate one - please let me know.

cc @astencel-sumo 

**Link to tracking Issue:** Please see comments in #31597. 

**Testing:** Tests were refactored to match the new config.

**Documentation:** Mostly copied from original receiver, some changes related to config occured.

### CODE REVIEW

File: .chloggen/docker-receiver.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [31597]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Currently, the receiver does the same things as Docker stats receiver. |
+         In the future, the receiver will be able to fetch also Docker events and container logs via Docker API.
+
```
Comment (aboguszewski-sumo) [inline .chloggen/docker-receiver.yaml:20]: Is changelog workflow failing because of incorrect usage of pipe here?

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -204,6 +204,7 @@ receiver/cloudfoundryreceiver/                           @open-telemetry/collect
 receiver/collectdreceiver/                               @open-telemetry/collector-contrib-approvers @atoulme
 receiver/couchdbreceiver/                                @open-telemetry/collector-contrib-approvers @djaglowski
 receiver/datadogreceiver/                                @open-telemetry/collector-contrib-approvers @boostchicken @gouthamve @jpkrohling @MovieStoreGuy
+receiver/dockerreceiver/                                 @open-telemetry/collector-contrib-approvers @rmfitzpatrick @jamesmoessis @aboguszewski-sumo
```
Comment (aboguszewski-sumo) [inline .github/CODEOWNERS:207]: The codeowners here are the codeowners of `dockerstats` receiver plus me.

File: receiver/dockerreceiver/config_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		Endpoint:         "someEndpoint",
+		DockerAPIVersion: "1.25",
+		Stats: StatsConfig{
+			ControllerConfig: scraperhelper.ControllerConfig{},
+		},
+	}
+	// The message is generated twice because when it's nested,
+	// it is returned both as an error of ControllerConfig and StatsConfig.
+	assert.Equal(t, `"collection_interval": requires positive value; "collection_interval": requires positive value`, component.ValidateConfig(cfg).Error())
+}
```
Comment (aboguszewski-sumo) [inline receiver/dockerreceiver/config_test.go:117]: Two errors are reported here. I checked with debugger and it seems that it's because `ControllerConfig` is nested inside `StatsConfig`.

File: .chloggen/docker-receiver.yaml (hunk 1 above)
Comment (astencel-sumo) [reply .chloggen/docker-receiver.yaml:20]: Do it like this: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1bb4967f33f26189e65a72f19b65fa50962ac698/.chloggen/add_serviceName_column_to_metrics_tables.yaml

```
subtext: |
  It's a breaking change. users who upgrade to the latest version need to alter the Clickhouse table:
  bla bla bla next line
  another line
```

File: extension/ackextension/go.mod [hunk 4]
```diff
@@ -6,27 +6,39 @@ require (
 	github.com/hashicorp/golang-lru/v2 v2.0.7
 	github.com/stretchr/testify v1.9.0
 	go.opentelemetry.io/collector/component v0.97.1-0.20240327181407-1038b67c85a0
+	go.opentelemetry.io/collector/confmap v0.97.1-0.20240327181407-1038b67c85a0
```
Comment (astencel-sumo) [inline extension/ackextension/go.mod:9]: Are changes in this file needed? Doesn't seem to be related to this pull request.

File: receiver/dockerreceiver/README.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+> :information_source: Requires Docker API version 1.22+ and only Linux is supported.
+
+## Configuration
+
+The following settings are optional and affect all telemetry:
+
+- `endpoint` (default = `unix:///var/run/docker.sock`): Address to reach the desired Docker daemon.
+- `api_version` (default = `1.25`): The Docker client API version (must be 1.25+). If using one with a terminating zero, input as a string to prevent undesired truncation (e.g. `"1.40"` instead of `1.40`, which is parsed as `1.4`). [Docker API versions](https://docs.docker.com/engine/api/).
+
+### Stats
```
Comment (astencel-sumo) [inline receiver/dockerreceiver/README.md:31]: I appreciate that you thought about separating metrics collection config from logs/events collection config at this early stage.

After looking at these config properties in detail, it seems to me that at least some of these could be useful for logs collection as well:
- `timeout` could probably easily be a common setting between both metrics and logs?
- `excluded_images` - I wonder if this setting could also allow to filter Docker events about containers? Probably need to check how this is implemented in the Docker client library.
- `container_labels_to_metrics_labes` - seems like this might be usable for Docker container events as well. Not sure about `env_vars_to_metric_labes`, probably need to check if env vars are available in the Docker events.

File: receiver/dockerreceiver/config.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/docker"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerreceiver/internal/metadata"
+)
+
+var _ component.Config = (*Config)(nil)
+
+type Config struct {
```
Comment (astencel-sumo) [inline receiver/dockerreceiver/config.go:18]: The approach I had in mind was for the common code for `docker_stats` receiver and the new `docker` receiver to not be duplicated, but instead extracted to a common package that would be used by both receivers. This is what we did for the Logging and Debug exporter here: https://github.com/open-telemetry/opentelemetry-collector/tree/v0.96.0/exporter/internal/common/. 

I understand that this is not as straightforward if the new `docker` receiver accepts configuration in different structure, but I think it's still doable (or maybe we can go with no changes to the structure of the config, as proposed in the [comment above](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32023/files#r1547655679)).

**Comment by astencel-sumo** [comment]
Also can the code owners of the original Docker Stats receiver please take a look :pray: @rmfitzpatrick @jamesmoessis

File: receiver/dockerreceiver/README.md (hunk 5 above)
Comment (aboguszewski-sumo) [reply receiver/dockerreceiver/README.md:31]: Let's consider the following user case: a user wants to collect metrics, excluding container `a` and events for all containers. What do we want them to do: create two separate `dockerreceiver`s or create one with different configs?

From my perspective, both are valid, but I'm not sure which one would be in the spirit of OpenTelemetry.

File: receiver/dockerreceiver/config.go (hunk 6 above)
Comment (aboguszewski-sumo) [reply receiver/dockerreceiver/config.go:18]: There already exists an `internal/docker` package, is it a good place for it or do we want to keep it in a separate package?

**Review by jamesmoessis — CHANGES_REQUESTED** [review]
Creating an entirely new receiver is IMO overkill. There are a lot of users of the docker stats receiver, so deprecating a component is going to need really good reasons. 

I can't see any reason given here or in the linked issue that needs this whole receiver to be copy-pasted under a different name. You want to add logs/events then I would suggest adding that as a feature to the existing receiver, or creating a separate receiver. If you think the name `dockerstats` is a problem then I would suggest looking at renaming it. 

The docker stats receiver is also still in alpha, so if you must make breaking changes to it, that's OK.

File: receiver/dockerreceiver/README.md (hunk 5 above)
Comment (astencel-sumo) [reply receiver/dockerreceiver/README.md:31]: This is a very good phrasing of the question :) I'm not aware of a definitive answer on "the OpenTelemetry way". I would lean towards the user specifying two different components for this. I'm definitely not in favor of having separate options in components that do (mostly) the same thing, but for a different signal: logs, metrics, traces.

File: receiver/dockerreceiver/config.go (hunk 6 above)
Comment (astencel-sumo) [reply receiver/dockerreceiver/config.go:18]: The `internal/docker` package defines the Docker client for use in other collector packages. I think having one package doing one thing makes it easier to work with a package.

The code we want to extract is concerned with code common to both Docker and Docker Stats receiver. How about package `internal/docker/receiver` or `receiver/internal/docker`?

**Comment by astencel-sumo** [comment]
> If you think the name `dockerstats` is a problem then I would suggest looking at renaming it.

This is exactly what we want to do here. See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31597#issuecomment-1985378982:

> Renaming a component in a graceful manner is a non-trivial endeavor. We did something similar with changing logging exporter to debug. How this would work in practice is perhaps:
> 
> 1. Create a new docker receiver that will share the metrics collection code with the existing docker_stats receiver. Deprecate the docker_stats receiver in favor of the docker receiver.
> 2. Add the new code for logs collection (Docker events, container logs - possibly in separate PRs) in the new docker receiver, leaving the old docker_stats receiver's functionality unchanged.
> 3. Remove the deprecated docker_stats receiver after a long time, perhaps 6 or 12 months.

We could indeed just rename the component and release a new version of the collector that does not have `docker_stats` receiver and instead has `docker` receiver. We are choosing a different path, that is harder for us - maintainers - but hopefully easier for users of the collector and also vendors of other distributions. We want to prevent unexpected breaking changes as much as possible, especially for popular components, even when the component is in "Alpha" stability.

By "we are choosing" I'm not sure whom I mean besides me :wink: So if there are other maintainers who think the simple rename procedure is fine, I'm open to changing it. I'll actually bring it to today's SIG meeting.

File: receiver/dockerreceiver/config.go (hunk 6 above)
Comment (aboguszewski-sumo) [reply receiver/dockerreceiver/config.go:18]: @andrzej-stencel after initial tries, I'm reluctant to this idea

1. It seems that even in case of simple components like `debug` and `logging` exporter code duplication is not easy to avoid: https://github.com/open-telemetry/opentelemetry-collector/commit/9731ea3281de3c14141c8ccd7fab0d14ee9aac73
2. In case of this (these) receiver(s), the subcomponent `metricsReceiver` has `Config` embedded inside. To avoid recursive imports, we would have to move `Config` to the common package, which makes little sense because we already want to change its specification. Currently, metric receiver's code is the most crucial part of the receiver, so it makes little sense to create a common package without it.

To sum up, it seems that the only advantage of moving this this way is making the reviewing process earlier. If that's what you wanted to achieve, then I can agree and do it. Otherwise I'd avoid creating a common subpackage.

File: receiver/dockerreceiver/README.md (hunk 5 above)
Comment (aboguszewski-sumo) [reply receiver/dockerreceiver/README.md]: @andrzej-stencel I made a PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33038
let's move further discussions on this aspect there.

**Comment by aboguszewski-sumo** [comment]
I'm closing this for now: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33038 has higher priority and I will recreate/reopen this PR once it gets accepted.
